pkg/router/controller/factory/factory.go

```diff
@@ -2,509 +2,265 @@ package factory
+	f.initInformers(rc, watchNodes, enableIngress)
+	f.initCallbacks(rc)
+	return rc
 }
 
-type routesByHost struct {
-	routes    cache.Indexer
-	endpoints cache.Store
-}
-
-func (r *routesByHost) Hosts() []string {
-	return r.routes.ListIndexFuncValues("host")
-}
+func (f *RouterControllerFactory) initInformers(rc *routercontroller.RouterController, watchNodes, enableIngress bool) {
+	f.createEndpointsSharedInformer(rc)
+	f.createRoutesSharedInformer(rc)
 
-func (r *routesByHost) Route(host string) (*routeapi.Route, bool) {
-	arr, err := r.routes.ByIndex("host", host)
-	if err != nil || len(arr) == 0 {
-		return nil, false
+	if watchNodes {
+		f.createNodesSharedInformer(rc)
 	}
-	return oldestRoute(arr), true
-}
 
-func (r *routesByHost) Endpoints(namespace, name string) *kapi.Endpoints {
-	obj, ok, err := r.endpoints.GetByKey(fmt.Sprintf("%s/%s", namespace, name))
-	if !ok || err != nil {
-		return &kapi.Endpoints{}
+	if enableIngress {
+		f.createIngressesSharedInformer(rc)
+		f.createSecretsSharedInformer(rc)
 	}
-	return obj.(*kapi.Endpoints)
 }
 
-// routeAge sorts routes from oldest to newest and is stable for all routes.
-type routeAge []routeapi.Route
-
-func (r routeAge) Len() int      { return len(r) }
-func (r routeAge) Swap(i, j int) { r[i], r[j] = r[j], r[i] }
-func (r routeAge) Less(i, j int) bool {
-	return routeapi.RouteLessThan(&r[i], &r[j])
-}
-
-func oldestRoute(routes []interface{}) *routeapi.Route {
-	var oldest *routeapi.Route
-	for i := range routes {
-		route := routes[i].(*routeapi.Route)
-		if oldest == nil || route.CreationTimestamp.Before(oldest.CreationTimestamp) {
-			oldest = route
+func (f *RouterControllerFactory) initCallbacks(rc *routercontroller.RouterController) {
+	rc.StartInformers = func(stopCh <-chan struct{}) {
+		for _, ref := range f.iRef {
+			go ref.informer.Run(stopCh)
 		}
 	}
-	return oldest
-}
 
-func hostIndexFunc(obj interface{}) ([]string, error) {
-	route := obj.(*routeapi.Route)
-	hosts := []string{
-		fmt.Sprintf("%s-%s%s", route.Name, route.Namespace, ".generated.local"),
-	}
-	if len(route.Spec.Host) > 0 {
-		hosts = append(hosts, route.Spec.Host)
+	rc.HasSyncedInformers = func() bool {
+		for _, ref := range f.iRef {
+			if !ref.informer.HasSynced() {
+				return false
+			}
+		}
+		return true
 	}
-	return hosts, nil
-}
 
-// routeLW is a ListWatcher for routes that can be filtered to a label, field, or
-// namespace.
-type routeLW struct {
-	client    routeclient.RoutesGetter
-	label     labels.Selector
-	field     fields.Selector
-	namespace string
-}
-
-func (lw *routeLW) List(options metav1.ListOptions) (runtime.Object, error) {
-	var label, field string
-	if lw.label != nil {
-		label = lw.label.String()
-	}
-	if lw.field != nil {
-		field = lw.field.String()
-	}
-	opts := metav1.ListOptions{
-		LabelSelector: label,
-		FieldSelector: field,
-	}
-	routes, err := lw.client.Routes(lw.namespace).List(opts)
-	if err != nil {
-		return nil, err
+	rc.FirstSyncDone = func() bool {
```

Comment:
Once has synced has returned true, this is unnecessary.

Maybe we should talk in person, I still think this is much more complicated than it needs to be.

1. Set "syncing" boolean true (never commit while this is true)
2. have your config loops update internal structs
3. wait for all informers synced
4. set syncing to be true
5. trigger a refresh in informers

3-5 should be able to be done in a single method.

Reply:
So the issue here is that you have two caches:

1. in the informer
2. caches in the route plugin

The second cache is filled by the first cache.  The code that fills the second cache is not done under the lock that maintains the first cache, and therefore while order of events can be guaranteed, there is no guarantee that the second cache is up to date with the first cache.

You need to determine when the second cache has observed all of the events from the first cache once it has synced once.  

I *think* you can address this by starting the informers, waiting for synced (on all of them) and then call each cache as informer.GetStore().List() and send those to the route controller.  Then call commit.  Then register the route controller, and every update from then on is safe.